controllers/imagerepository_notifications.go

```diff
@@ -0,0 +1,173 @@
+		// No status notifications to check
+		return nil
+	}
+	allNotifications, err := r.QuayClient.GetNotifications(r.QuayOrganization, imageRepository.Spec.Image.Name)
+	if err != nil {
+		if err = r.UpdateImageRepositoryStatusMessage(ctx, imageRepository, "Couldn't retrieve all Quay notifications"); err != nil {
+			return err
+		}
+		return nil
+	}
+	for index := 0; index < len(imageRepository.Status.Notifications); index++ {
+		statusNotification := &imageRepository.Status.Notifications[index] // This way we can update the UUID if notification gets updated
+		existsInSpec := false
+		for _, notification := range imageRepository.Spec.Notifications {
+			if notification.Title == statusNotification.Title {
+				existsInSpec = true
+				quayNotification := notificationExistsInQuayByUUID(statusNotification.UUID, imageRepository, allNotifications)
+				if isNotificationChanged(notification, quayNotification) {
+					// Updated item in Spec.Notifications: update notification in Quay
+					log.Info("Updating notification in Quay", "Title", notification.Title, "Event", notification.Event, "Method", notification.Method, "UUID", quayNotification.UUID)
+					updatedNotification, err := r.QuayClient.UpdateNotification(
+						r.QuayOrganization,
+						imageRepository.Spec.Image.Name,
+						statusNotification.UUID,
+						quay.Notification{
+							Title:  notification.Title,
+							Event:  string(notification.Event),
+							Method: string(notification.Method),
+							Config: quay.NotificationConfig{
+								Url: notification.Config.Url,
+							},
+							EventConfig: quay.NotificationEventConfig{},
+						})
+					if err != nil {
+						log.Error(err, "failed to update notification", "Title", statusNotification.Title, "UUID", statusNotification.UUID)
+						if err = r.UpdateImageRepositoryStatusMessage(ctx, imageRepository, "Error while updating the notification in Quay"); err != nil {
+							return err
+						}
+						return nil
+					}
+					statusNotification.UUID = updatedNotification.UUID
+				}
+			} else {
+				quayNotification := notificationExistsInQuayByTitle(notification.Title, imageRepository, allNotifications)
+				if quayNotification.UUID == "" {
+					// New item in Spec.Notifications: add notification to Quay and Status.Notifications
+					log.Info("Adding new notification to Quay", "Title", notification.Title, "Event", notification.Event, "Method", notification.Method)
+					notificationStatus, err := r.AddNotification(notification, imageRepository)
+					if err != nil {
+						log.Error(err, "failed to add notification", "Title", notification.Title, "Event", notification.Event, "Method", notification.Method)
+						if err = r.UpdateImageRepositoryStatusMessage(ctx, imageRepository, "Error while adding a notification to Quay"); err != nil {
+							return err
+						}
+						return nil
+					}
+					imageRepository.Status.Notifications = append(imageRepository.Status.Notifications, notificationStatus)
```

Comment:
You update the notification array while iterating over it in the `for` loop. I am not familiar with how Golang handles this situation but other languages usually end up in an unpredictable state. Is this operation safe in golang?

Reply:
Yes, I took that into consideration. I verified it woks by using a system like the one in L49 to update some fields and to reflect the status updates it works by updating the ImageRepository status with `r.Client.Status().Update(ctx, imageRepository)` calls. I noticed that same strategy is followed in different parts of the codebase.

Reply:
At the beginning the status update call was placed inside `HandleNotifications` but after Mykola's suggestions I moved it right after calling the function in the reconcile loop.

Reply:
this is skipping 1 element of imageRepository.Status.Notifications, because you remove one element, the rest is moved and now that 1st element which should be processed is considered processed 

verified with simple code:

	ar = []int{1, 2, 3, 4, 5, 6}
	fmt.Println(ar)
	for index := 0; index < len(ar); index++ {
		element := &ar[index]
		fmt.Println(*element)
		if index == 1 {
			ar = append(ar[:index], ar[index+1:]...)
		}
	}


output is :

[1 2 3 4 5 6]
1
2
4
5
6

so '3' is skipped because it has now (after append) index which was already processed

Reply:
You're right. I didn't modify the index at the same time (`index--`). Thanks for pointing this out.